777advait/precommit hooks

777advait opened the pull request:
fixes #126

HarshKapadia2 commented:
Hello, @777advait, thank you so much for your interest and for your PR!

I had a few questions and tasks
- Why is the `bun.lockb` file required?
- When does the Husky linting command run? It didn't run for me.
- When linting, there should be a linting check (no file modification) and if there are errors, then the commit should not be allowed. Please add this logic if possible.
- The issue this PR means to fix (#126) also mentions syncing the linting check to GitHub Actions. Can you please do that?
- Can you please add the word 'update' to the 'type' part of the conventional commit format for commitlint?
- Can you please add documentation to `CONTRIBUTING.md` file on the entire pre-commit hook process?
	- What hooks exist, when they run, how to run them manually and any other things to note, like the commit format, etc.
- The `package.json` file has been converted from tab indentations to space indentations. Please reverse that. We want to use tab indentations wherever possible.
	- The linting pre-commit hooks should've caught that, as tabbing being preferred and the tab width are both mentioned in `.prettierrc`.

777advait commented:
Hey @HarshKapadia2, i've made some new commits, pls take a look and lmk if its good i'll get started with updating `CONTRIBUTING.md` file then (you can change the status of this PR to draft until then). also in the latest commits there were some linting issues in adoc files (there were just too many) so i've pushed without running the lint check

> Why is the bun.lockb file required?

its the lock file for bun package manager, we can have it if anyone else uses bun too, otherwise i'll untrack it

> When does the Husky linting command run? It didn't run for me.

i mistakenly had the lint script in a separate file i've fixed it now

> When linting, there should be a linting check (no file modification) and if there are errors, then the commit should not be allowed. Please add this logic if possible.

changed from `npm run lint` to `npm run lint:check` ig that should do the work

> The issue this PR means to fix (#126) also mentions syncing the linting check to GitHub Actions. Can you please do that?

<img width="318" alt="image" src="https://github.com/user-attachments/assets/32d4fb79-c3bb-4c08-8e29-f37495755574" />

ig we're already checking linting in the action, could you elaborate if thats not what you meant?

> Can you please add the word 'update' to the 'type' part of the conventional commit format for commitlint?

updated in the latest commit, you can update the `rules.types-enum` object in `commitlint.config.js` to add more custom types

> The package.json file has been converted from tab indentations to space indentations. Please reverse that. We want to use tab indentations wherever possible.

reverted in the latest commit

HarshKapadia2 commented on .husky/pre-commit:
```diff
@@ -0,0 +1,7 @@
+npm run commitlint ${1}
+
+echo "🛠️ Running lint check..."
+npm run lint:check
```
```suggestion
npm run lint:check

```

Please add a new line at the end of the file.

HarshKapadia2 commented on package.json:
```diff
@@ -7,7 +7,9 @@
 		"lint:check": "prettier --check .",
-		"lint": "prettier --write ."
+		"lint": "prettier --write .",
+		"prepare": "husky || true",
```
Who is using the `prepare` command and when is it called?

KartikSoneji replied:
Husky does: https://typicode.github.io/husky/how-to.html#manual-setup
Note that `true` doesn't exist by default on Windows so if `husky` fails then `prepare` will fail too.

777advait replied:
https://typicode.github.io/husky/how-to.html#ci-server-and-docker
the `|| true` part is for CI environments so shouldn't be an issue

HarshKapadia2 commented on package.json:
```diff
@@ -29,9 +31,13 @@
 		"typescript": "^5.3.3"
 	}
-}
+}
```
```suggestion
}

```

Please add a new line at the end of the file.

KartikSoneji replied:
We should add an `.editorconfig` file for this.
https://editorconfig.org/#overview

HarshKapadia2 commented:
> also in the latest commits there were some linting issues in adoc files (there were just too many) so i've pushed without running the lint check

Almost every summary is affected. @KartikSoneji can you please help us resolve the issues?

> > Why is the bun.lockb file required?
>
> its the lock file for bun package manager, we can have it if anyone else uses bun too, otherwise i'll untrack it

We don't need it because all our work is in Node.js. Please remove the file.

> > When does the Husky linting command run? It didn't run for me.
>
> i mistakenly had the lint script in a separate file i've fixed it now

Thank you. Let's lint the files before checking the commit format.

> > The issue this PR means to fix (https://github.com/OurTechCommunity/catchup/issues/126) also mentions syncing the linting check to GitHub Actions. Can you please do that?
>
> ig we're already checking linting in the action, could you elaborate if thats not what you meant?

Oh yes, you're correct. Please let us know if you wanted anything more than what currently exists in the Actions file, @KartikSoneji.

> > When linting, there should be a linting check (no file modification) and if there are errors, then the commit should not be allowed. Please add this logic if possible.
>
> changed from npm run lint to npm run lint:check ig that should do the work

Yes, works. Thank you!

> > Can you please add the word 'update' to the 'type' part of the conventional commit format for commitlint?
>
> updated in the latest commit, you can update the rules.types-enum object in `commitlint.config.js` to add more custom types

I cannot seem to get rid of a previous commit message (`testing-2`) as the commitlint input even after starting a new terminal...

<img width="652" alt="{6A9E266E-34BA-4A89-B690-33B0B9C89994}" src="https://github.com/user-attachments/assets/7d5bc378-fb00-4255-9ede-de17ad61858e" />

KartikSoneji commented:
Hmm the adoc plugin might not be stable enough for us to use at the moment.

777advait commented:
@HarshKapadia2 added empty lines at EOF as mentioned and added `*.adoc` to .prettierignore for now. also check the documentation for pre-commit hooks

> I cannot seem to get rid of a previous commit message (testing-2) as the commitlint input even after starting a new terminal...

you seem to have committed changes with the commit message `testing-2` in the git log, i believe soft resetting the log and committing with appropriate commit message format will fix the issue

777advait commented:
also added a standalone husky install script, so it wont have an issue on windows environments now

HarshKapadia2 commented:
Hey @777advait, thank you so much for all the effort! Appreciate you working with us and having all these back-and-forths!

I pushed a commit with quite a few changes. Please go through all the changes @777advait.

- Commitlint wants a `commit-msg` hook, and not a `pre-commit` hook. ([1](https://commitlint.js.org/guides/local-setup.html#:~:text=To%20use%20commitlint,commit%2Dmsg%20hook.), [2](https://github.com/conventional-changelog/commitlint/issues/589#issuecomment-1251895709))
- A `bun.lockb` file should just not be mentioned anywhere in the project. Our project has nothing to do with Bun. (Removed the mention from `.gitignore`.)
- The `CONTRIBUTING.md` file needed some grammar and formatting updates.
	- @777advait, it is always good to try to modify one's writing style based on the rest of the file or to the entire project's style. I would recommend going through the changes I made in this file.
- Added an `.editorconfig` file as per [Kartik's recommendation](https://github.com/OurTechCommunity/catchup/pull/268#discussion_r2094630790).
- Changed the `commitlint` script in `package.json`
- Committed the updated `package-lock.json` file.
- The one file I'm not sure about is `.husky/install.mjs`.
	- @KartikSoneji can you please comment on whether the `process.env` variables are actually available in CI environments?
	- Also, @KartikSoneji, do we even need to mention a skip for Husky installation in production environments? Dev dependencies are not installed in production environments, right?
	- @KartikSoneji should we even mention a skipping condition for Husky? So what if it is installed during CI? Will it cause issues with Netlify's deploys? I don't think so, but I just want to be sure, because I'm not sure what they do. (I'm not mentioning production environments here because I'm assuming dev dependencies are just not present in production environments.)

HarshKapadia2 commented:
Hey @KartikSoneji, can you please finalize this PR and merge it?

HarshKapadia2 commented:
> > should we even mention a skipping condition for Husky?
>
> I think yes, since it can get tedious to run on each local commit.

In the current format, the way to skip all Husky hooks would be to use the `--no-verify` flag with the `git commit` command. This has been documented in the `CONTRIBUTING.md` file.

```bash
$ git commit --no-verify -m "test: commit_msg_here"
```

---

From your messages, I think we're good to merge this.

Thank you so much for all your efforts, @777advait! We appreciate you! If you're interested in helping us further, you can have a look at https://github.com/OurTechCommunity/catchup/issues/184. It already has a PR that probably only needs modification and updating. Let us know if you would want to take that up.
